Documentation/teaching/lectures/interrupts.rst

```diff
@@ -0,0 +1,558 @@
+a system call are examples of exceptions.
+
+Asynchronous interrupts, usually named interrupts, are external events generated
+by I/O devices. For example a network card generates an interrupts to signal
+that a packet has arrived.
+
+Exceptions
+----------
+
+There are two sources for exceptions:
+
+.. slide:: Exceptions
+   :inline-contents: True
+   :level: 2
+
+   * processor detected
+
+     - **faults**
+     - **traps**
+     - **aborts**
+   * programmed
+
+
+     - **int n**
+
+Processor detected exceptions are raised when an abornmal condition is
+detected while executing an instruction.
+
+A fault is a type of exception that is reported before the execution of the
+instruction and can be usually corrected. The saved EIP is the address of
+the instruction that caused the fault, so after the fault is corrected
+the program can re-execute the faulty instruction. (e.g page fault).
+
+A trap is a type of exception that is reported after the execution of the
+instruction in which the exception was detected. The saved EIP is the address
+of the instruction after the instuction that caused the trap. (e.g debug trap).
+
+Hardware
+========
+
+Programmable Interrupt Controller
+---------------------------------
+
+.. slide:: Hardware (PIC)
+   :inline-contents: True
+   :level: 2
+
+   .. ditaa::
+        
+        +-----------+   NMI
+        |           |
+        |           <----------+
+        |           |
+        |           |           +------------+
+        |           |           |            |   IRQ0
+        |           |           |            <------------+ device0
+        |    CPU    |           |            |   IRQ1
+        |           |   INTR    |     PIC    <------------+ device1
+        |           <-----------+            |   IRQN
+        |           |           |            <------------+ deviceN
+        |           |           |            |
+        +-----------+           +------------+
+        
+A device supporting interrupts has an output pin used for signalling an Interrupt ReQuest. IRQ
+pins are connected to a device named Programmable Interrupt Controller (PIC) which is connected
+to CPU's INTR pin.
+
+A PIC usually has a set of ports used to exchange information with the CPU. When a device
+connected to one of the PIC's IRQ lines needs CPU attention the following flow happens:
+   * device raises an interrupt on the corresponding IRQn pin
+   * PIC converts the IRQ into a vector number and writes it to a port for CPU to read
+   * PIC raises an interrupt on CPU INTR pin
+   * PIC waits for CPU to acknowledge an interrupt
+   * CPU handles the interrupt
+
+Will see later how the CPU handles the interrupt. Important to notice is that by design PIC won't raise
+another interrupt until the CPU acknowledged the current interrupt.
+
+Each IRQ line can be individually disabled. This allows simplifying design by making sure that
+interrupt handlers are always executed serially.
+
```

Comment:
Yes, this is a good idea. 

I have some questions on this topic. 

I know most of this is hardware dependent but I want to make sure I got the correct idea.

1) Lets say a PIC has 8 IRQ lines. What happens at PIC level when IRQ0 is asserted? PIC tries to notify the CPU via INTR pin, in this time any interrupt arriving on IRQ0...7 is ignored until the PIC ACK's IRQ0. 

I assume that the interrupts arriving until the PIC ACK's IRQ0 are not lost but remembered some how and delivered after PIC gets the ACK from CPU.

2) cli/sti - disable/enable the interrupts only on local CPU? Or this is global? If a maskable interrupt arrives while it is masked the kernel still keeps track of it and delivers it when interrupts are unmasked,right?

3) disable_irq(irq), disables irq globally. What happens if an interrupts arrives while it was disabled? The core kernel code will still deliver it (by calling the irq handler) later?

Now, the following slide says that there are 3 phases in the process
of handling an interrupt:

https://linux-kernel-labs.github.io/pull/30/lectures/interrupts-slides.html#15

a) critical, i think this is the PIC level handling when all IRQ lines are disabled.
b) immediate, i think this is the CPU level handling only when the given IRQ line is disabled.
c) deferred, all IRQ are enabled.

Another set of questions: 

1) While an interrupt handler runs as a result of IRQ0, is that handler running only with IRQ0
disabled locally(?)/globally(?) but with the rest of IRQ enabled?
 
This means that when an interrupt handler runs, it can be interrupted by another interrupt but not by the same interrupt.

Reply:
> Lets say a PIC has 8 IRQ lines. What happens at PIC level when IRQ0 is asserted? PIC tries to notify the CPU via INTR pin, in this time any interrupt arriving on IRQ0...7 is ignored until the PIC ACK's IRQ0.

Any new interrupts are "buffered" by the PIC until it gets an ACK from the CPU. 

>cli/sti - disable/enable the interrupts only on local CPU? Or this is global? If a maskable interrupt arrives while it is masked the kernel still keeps track of it and delivers it when interrupts are unmasked,right?

cli/sti is per CPU/core. Yes, the interrupt will be "buffered" until STI when the interrupt will be triggered.

>disable_irq(irq), disables irq globally. What happens if an interrupts arrives while it was disabled? The core kernel code will still deliver it (by calling the irq handler) later?

Yes, it is disabled globally since it is done at the PIC level. disable_irq() makes sure there are no pending interrupts after the function returns. There is a disable_irq_nosync() which does not guarantee this.

> Now, the following slide says that there are 3 phases in the process
of handling an interrupt:
> https://linux-kernel-labs.github.io/pull/30/lectures/interrupts-slides.html#15
> a) critical, i think this is the PIC level handling when all IRQ lines are disabled.
> b) immediate, i think this is the CPU level handling only when the given IRQ line is disabled.
> c) deferred, all IRQ are enabled.

For both a) and b) local CPU interrupts (IF flag) are disabled. Before e58aa3d2d0cc01ad8d6f7f640a0670433f794922("genirq: Run irq handlers with interrupts disabled") b) was running with CPU interrupts enabled.

The individual IRQ line is most likely (depends on the arch and PIC) still enabled during a) and it may get disabled during a) (e.g. for a level edge interrupt) or it may get disabled during b) (e.g. network drivers using NAPI).

>  While an interrupt handler runs as a result of IRQ0, is that handler running only with IRQ0
disabled locally(?)/globally(?) but with the rest of IRQ enabled?

See above.